hydra-main/src/main/java/com/addthis/hydra/query/aggregate/QueryTask.java

```diff
@@ -55,54 +59,13 @@ public void run() {
                 sourceAggregator.needScheduling = true;
                 return;
             }
-            QueryTaskSource[] taskSources = sourceAggregator.taskSources;
-            int bundlesProcessed = 0;
-            int complete = 0;
-            boolean processedBundle = true;
-            while (processedBundle && (bundlesProcessed < AggregateConfig.FRAME_READER_READS)) {
-                complete = 0;
-                processedBundle = false;
-                for (int i = 0; i < taskSources.length; i++) {
-                    QueryTaskSource taskSource = taskSources[i];
-                    if (taskSource.complete()) {
-                        complete += 1;
-                        continue;
-                    }
-                    try {
-                        Bundle nextBundle = taskSource.next();
-                        if (nextBundle != null) {
-                            sourceField.setValue(nextBundle,
-                                                 create(taskSource.getSelectedSource().queryReference.name));
-                            sourceAggregator.consumer.send(nextBundle);
-                            processedBundle = true;
-                            bundlesProcessed += 1;
-                        } else if (!taskSource.oneHasResponded()
-                                   && taskSource.hasNoActiveSources()
-                                   && !sourceAggregator.queryPromise.isDone()) {
-                            log.debug("query task has no active options; attempting to lease one");
-                            for (QueryTaskSourceOption option : taskSource.options) {
-                                if (sourceAggregator.tryActivateSource(option)) {
-                                    log.debug("task option leased and activated successfully");
-                                    break;
-                                }
-                            }
-                        }
-                    } catch (IOException io) {
-                        if (taskSource.lines == 0) {
-                            // This QuerySource does not have this file anymore. Signal to the caller that a retry may resolve the issue.
-                            sourceAggregator.replaceQuerySource(taskSource, taskSource.getSelectedSource());
-                        } else {
-                            // This query source has started sending lines. Need to fail the query.
-                            throw io;
-                        }
-                    }
-                }
-            }
+            // NOTE: both provider and readFrame update sourceAggregator.completed
+            List<QueryTaskSource> taskSources = taskSourceProvider.get();
```

Comment:
I'm not sure how exactly "iterating over the array" should work? One option is to make `list` and `inactive` field variables. The `get` method will update their elements, instead of creating them every time.